lib/streamlit/elements/file_uploader.py

```diff
@@ -167,10 +167,15 @@ def serialize_file_uploader(
                 ids = [f.id for f in files]
             else:
                 ids = [files.id]
+
             ctx = get_report_ctx()
             if ctx is None:
                 return []
-            max_id = ctx.uploaded_file_mgr._file_id_counter
+
+            # ctx.uploaded_file_mgr._file_id_counter stores the id to use for
+            # the next uploaded file, so the current highest file id is the
+            # counter minus 1.
+            max_id = ctx.uploaded_file_mgr._file_id_counter - 1
```

Comment:
Is there a test that would've caught this off-by-one?

Reply:
One of the revived file_uploader unit tests now verifies that the counter is decremented as it should be by the time the list of file_ids makes it to the call to `remove_orphaned_files`. 

This feels a bit unsatisfactory to me, but I wasn't able to figure out a way to more thoroughly test this.